### ISSUE TITLE
Fix erroneously double incremented and decremented consumers

### ISSUE THREAD

**Issue by liamaharon** [paritytech/polkadot-sdk#2037]
Needs to be implemented once https://github.com/paritytech/polkadot-sdk/pull/1976 is merged and released.

Two options have been floated:
1. write a script that scrapes extrinsics for accounts that had their consumers erroneously incremented and generates a migration to decrement them
2. introduce `fn fix_consumers(origin, who: AccountId)` in the System pallet, with an additional `frame_system::Config::CountConsumers` trait with a function `fn count_consumers(who: &AccountId) -> u32`, which could possibly be written on a per-pallet basis and composed in a tuple and used by AllPalletsWithSystem`.

**Comment by xlc** [comment]
2 will be useful and we can easily convert it into a try_runtime invariant to ensure the consistency in future

**Comment by Ank4n** [comment]
Has there been any progress/update to this? We have recently found [pools are failing to be destroyed](https://github.com/paritytech/polkadot-sdk/issues/4440) because of this bug.

**Comment by ggwpez** [comment]
I think that it would help to have https://github.com/paritytech/polkadot-sdk/pull/4398 deployed before we start starting migrations for the references...

**Comment by ggwpez** [comment]
I think there is no clear forward yet. I guess the `count_consumer` approach could work, but its a bit complicated in general.

**Comment by liamaharon** [comment]
I think the way forward is to scrape bad `update_locks` invocations (https://github.com/paritytech/polkadot-sdk/issues/1970), and write a migration to fix consumers based on that. A bit fiddly, but should be feasible I think?

**Comment by ggwpez** [comment]
Could work, but we have this issue not just on Polkadot - but presumably on all chains...  
What you mean is the go through an archive and record and erroneous double-decrements?

**Comment by liamaharon** [comment]
Yes, write a script that replays transactions and checks after every lock/unlock whether there was an erroneous increment/decrement. Then it crafts a migration which can be inserted into the runtime to fix them.

**Comment by Juanma0x** [comment]
Is there any news about allowing currently 'destroying' pools to be destroyed?

**Comment by ggwpez** [comment]
> Is there any news about allowing currently 'destroying' pools to be destroyed?

https://github.com/paritytech/polkadot-sdk/pull/4503 i suppose. Should be in the SDK 1.13 release.

**Comment by Juanma0x** [comment]
Pools can be destroyed now. I believe this issue can be closed.

Example: 14E8ZGhchDwhB3igyEudWJ5j3gPGKwT5ki3yStUy3XbgdkwX

**Comment by ggwpez** [comment]
There was a fix for pools: https://github.com/paritytech/polkadot-sdk/pull/4503 but the general issue is not addressed by this.

**Comment by Juanma0x** [comment]
I'm curious to know how progress on this issue is coming along. We've had several users reach out about being unable to empty their accounts fully, and I'd like to provide them with an update. Any information would be helpful!

**Comment by ggwpez** [comment]
>We've had several users reach out about being unable to empty their accounts fully

The only thing remaining is the ED? There is no straight forward fix for this historic issue. Fixing this properly would require quite some effort - i think more than we currently have available. Maybe @kianenigma has some ideas.

**Comment by kianenigma** [comment]
> I'm curious to know how progress on this issue is coming along. We've had several users reach out about being unable to empty their accounts fully, and I'd like to provide them with an update. Any information would be helpful!

Can you give us an idea of how many users will be affected by this, and what is the total amount of DOTs that will be "locked" while this is unfixed? 

You can probably do this by writing a script that scans all accounts, and checks those that cannot be killed now. 

This will help us prioritize this properly.

**Comment by Juanma0x** [comment]
I'm more than happy to communicate the solution to affected users, but locating every affected account may not fall entirely within my scope. But let me know if I can help in any other way.

**Comment by ggwpez** [comment]
>We've had several users reach out about being unable to empty their accounts fully

Does the existential deposit stay in there or what is the error when they try to use `transfer_all`?

**Comment by Juanma0x** [comment]
> > We've had several users reach out about being unable to empty their accounts fully
> 
> Does the existential deposit stay in there or what is the error when they try to use `transfer_all`?

The message displayed by Polkadot-JS UI when the user tries to reap the account is "There is an existing reference count on the sender account. As such the account cannot be reaped from the state." 

The ED stays after an affected account issues the `transfer_all` extrinsic (e.g., [24208885-2](https://polkadot.subscan.io/extrinsic/24208885-2)).

`transfer_allow_death` fails, returning a `token:Frozen` error (e.g., [24145806-2](https://polkadot.subscan.io/extrinsic/24145806-2)).

**Comment by alko89** [comment]
Just got sent here by support, seems one of my Ledger accounts is affected by this issue. Making a tx, from Ledger app, instead from Polkadot UI, makes the transaction being sent, but failing

**Comment by AlbertW3F** [comment]
> We can fix this as part of the Asset Hub migration ([#7632](https://github.com/paritytech/polkadot-sdk/issues/7632)) by counting the refs and then setting them correctly again. cc [@muharem](https://github.com/muharem)

Is it expected that this issue will be fixed during the Polkadot Asset Hub migration, as mentioned here?
